src/storage/src/storage/upload_source.rs

```diff
@@ -205,23 +214,84 @@ impl Seek for BytesSource {
     }
 }
 
+pub struct IterSource {
+    contents: Vec<bytes::Bytes>,
+    current: VecDeque<bytes::Bytes>,
+}
+
+impl IterSource {
+    pub(crate) fn new<I>(iterator: I) -> Self
+    where
+        I: IntoIterator<Item = bytes::Bytes>,
+    {
+        let contents: Vec<bytes::Bytes> = iterator.into_iter().collect();
+        let current: VecDeque<bytes::Bytes> = contents.iter().cloned().collect();
+        Self { contents, current }
+    }
+}
+
+impl StreamingSource for IterSource {
+    type Error = std::io::Error;
+
+    async fn next(&mut self) -> Option<std::result::Result<bytes::Bytes, Self::Error>> {
+        self.current.pop_front().map(Ok)
+    }
+
+    fn size_hint(&self) -> (u64, Option<u64>) {
+        let s = self.contents.iter().fold(0_u64, |a, i| a + i.len() as u64);
+        (s, Some(s))
+    }
+}
+
+impl Seek for IterSource {
+    type Error = std::io::Error;
+    async fn seek(&mut self, offset: u64) -> std::result::Result<(), Self::Error> {
+        let mut current: VecDeque<_> = self.contents.iter().cloned().collect();
```

Comment:
if this is a public type, this implementation should be more efficient. We only need to clone the chunks from `contents` which have already been consumed from `current`.